Adding recipe for TREXIO library

q-posev opened the pull request:
Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [X] If static libraries are linked in, the license of the static library is packaged.
- [X] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [X] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/trexio```) and found it was in an excellent condition.

q-posev commented:
@conda-forge/staged-recipes This PR is ready for review, thank you in advance!

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I was asked to ping @conda-forge/staged-recipes and so here I am doing that.

carterbox commented on recipes/trexio/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+  home: https://github.com/TREX-CoE/trexio
+  license: BSD-3-Clause
+  license_family: BSD 
+  license_file: COPYING
```
```suggestion
  license_file: LICENSE
```
Copying is a symbolic link to LICENSE. Add LICENSE directly?

q-posev replied:
Right, good idea! Done.

q-posev replied:
In fact, I just realized that `COPYING` is a symbolic link only when building from the GitHub repo clone. In the distribution tarball produced by Autotools (`make dist`) it appears as a text file, so it's fine. The `LICENSE` is not shipped with the distribution tarball. So we can keep it for now.

carterbox commented on recipes/trexio/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+
+test:
+  commands:
+    - test -f $PREFIX/lib/libtrexio.a  # [unix]
```
Please check for all the files that you are shipping. You should be shipping a `shared` library and header files instead of a static library with no headers.

q-posev replied:
Done.

carterbox commented on recipes/trexio/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+
+  host:
+    - hdf5
+    # the dependencies below are not required by TREXIO but they get exported from HDF5
+    # and if they are not listed here than the `conda-build --error-overlinking` fails
+    - zlib    
+    - openssl 
+    - libcurl 
```
```
  ERROR (trexio,lib/libtrexio.so.0.0.0): Needed DSO lib/libcurl.so.4 found in ['libcurl']
  ERROR (trexio,lib/libtrexio.so.0.0.0): .. but ['libcurl'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
```
This error message from conda means that it has inspected this binary and found that it links to libcurl (like when you inspect a file with ldd). Since, you state in this comment that libcurl is not a dependency of libtrexio, you should double check your build process to prevent overlinking.

You should *not* add the packages to the host section if trexio doesn't need them. If they are needed by hdf5 at runtime, hdf5 will have a run requirement for them.

q-posev replied:
Indeed, hdf5 has a run requirement for these packages (see their [meta.yaml](https://github.com/conda-forge/hdf5-feedstock/blob/master/recipe/meta.yaml)). Our build system based on Autotools parses the libraries that are linked against HDF5 and adds them to the `LIBS` list, which results in overlinking. I switched to the CMake build and there is no overlinking issue. Probably the `FindHDF5.cmake` module was designed with these things in mind.

carterbox commented on recipes/trexio/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - openssl 
+    - libcurl 
+
+  run:
+    - hdf5 
```
This shouldn't be required because hdf5 has a run export for itself.

q-posev replied:
Done. 

Here I had one question: since `trexio` depends on `hdf5`, should this dependency be somehow exported as a runtime dependency? Or this is taken into account by `conda` when listing `hdf5` under `requirements/host`?

carterbox replied:
Because the recipe for hdf5 contains a run export, conda-build automatically adds hdf5 to the `requirements/run` section of your recipe when you add it to the `requirements/host` section.

carterbox reviewed (changes requested):
The README on the dev repository says that CMake is for persons building from the tarball. Is that a strict requirement or do you just prefer autotools?

For autotools, the README includes calls to autogen.sh

q-posev commented:
> The README on the dev repository says that CMake is for persons building from the tarball. Is that a strict requirement or do you just prefer autotools?

We actually support both builds, it is up to the user to decide whether he would like to use CMake or autotools. Would you recommend to use CMake when building for `conda` ?

xhochy commented:
CMake is preferable to autotools because we have a (slightly) better infrastructure set up for it. But also take into account that there sometimes differences between those builds: https://conda-forge.org/docs/maintainer/knowledge_base.html#moving-from-an-autotools-build-to-a-cmake-build

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/trexio```) and found some lint.

Here's what I've got...


For **recipes/trexio**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [60]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/trexio```) and found it was in an excellent condition.

q-posev commented:
@carterbox I think this PR is ready for a review. Changes in a nutshell:

- Switched to CMake build to prevent overlinking issue
- Added MPI variant
- Removed  hdf5-related `host` and `run` requirements
- Added `run_exports` for trexio itself

carterbox commented on recipes/trexio/meta.yaml:
```diff
@@ -0,0 +1,83 @@
+test:
+  commands:
+    - test -f ${PREFIX}/lib/libtrexio${SHLIB_EXT}  # [unix]
+    - test -f ${PREFIX}/include/trexio.h  # [unix]
```
```suggestion
    - test -f ${PREFIX}/include/trexio.h  # [unix]
    - test -f ${PREFIX}/include/trexio_f.f90  # [unix]
```
